File: CleverTapSDK/Encryption/CTAESCrypt.m
//
//  CTAESCrypt.m
//  CleverTapSDK
//
//  Created by Kushagra Mishra on 07/02/25.
//  Copyright © 2025 CleverTap. All rights reserved.
//

#import "CTAESCrypt.h"
#import "CTConstants.h"

@implementation CTAESCrypt

- (nullable NSData *)AES128WithOperation:(CCOperation)operation
                                   key:(NSString *)key
                            identifier:(NSString *)identifier
                                  data:(NSData *)data {
    // Note: The key will be 0's but we intentionally are keeping it this way to maintain
    // compatibility. The correct code is:
    // char keyPtr[[key length] + 1];
    char keyCString[kCCKeySizeAES128 + 1];
    memset(keyCString, 0, sizeof(keyCString));
    [key getCString:keyCString maxLength:sizeof(keyCString) encoding:NSUTF8StringEncoding];
    
    char identifierCString[kCCBlockSizeAES128 + 1];
    memset(identifierCString, 0, sizeof(identifierCString));
    [identifier getCString:identifierCString
                 maxLength:sizeof(identifierCString)
                  encoding:NSUTF8StringEncoding];
    
    size_t outputAvailableSize = [data length] + kCCBlockSizeAES128;
    void *output = malloc(outputAvailableSize);
    
    size_t outputMovedSize = 0;
    CCCryptorStatus cryptStatus = CCCrypt(operation,
                                        kCCAlgorithmAES128,
                                        kCCOptionPKCS7Padding,
                                        keyCString,
                                        kCCBlockSizeAES128,
                                        identifierCString,
                                        [data bytes],
                                        [data length],
                                        output,
                                        outputAvailableSize,
                                        &outputMovedSize);
    
    if (cryptStatus != kCCSuccess) {
        CleverTapLogStaticInternal(@"Failed to encode/deocde the string with error code: %d", cryptStatus);
        free(output);
        return nil;
    }
    
    return [NSData dataWithBytesNoCopy:output length:outputMovedSize];
}

- (NSString *)generateKeyPassword {
    // Add your implementation here
    // Return the key password string
    return @"key";
}

@end

File: CleverTapSDK/Encryption/CTAESCrypt.swift
import Foundation
import CommonCrypto

class CryptoUtils {
    // AES128 encryption/decryption method
    func aes128(operation: CCOperation,
                key: String,
                identifier: String,
                data: Data) -> Data? {
        // Prepare key
        var keyCString = [CChar](repeating: 0, count: kCCKeySizeAES128 + 1)
        key.getCString(&keyCString, maxLength: keyCString.count, encoding: .utf8)
        
        // Prepare identifier
        var identifierCString = [CChar](repeating: 0, count: kCCBlockSizeAES128 + 1)
        identifier.getCString(&identifierCString, maxLength: identifierCString.count, encoding: .utf8)
        
        // Prepare output buffer
        let outputAvailableSize = data.count + kCCBlockSizeAES128
        let outputBuffer = UnsafeMutableRawPointer.allocate(byteCount: outputAvailableSize, alignment: 1)
        defer { outputBuffer.deallocate() }
        
        var outputMovedSize: size_t = 0
        
        let cryptStatus = data.withUnsafeBytes { dataBytes in
            CCCrypt(
                operation,
                CCAlgorithm(kCCAlgorithmAES),
                CCOptions(kCCOptionPKCS7Padding),
                keyCString,
                kCCBlockSizeAES128,
                identifierCString,
                dataBytes.baseAddress,
                data.count,
                outputBuffer,
                outputAvailableSize,
                &outputMovedSize
            )
        }
        
        guard cryptStatus == kCCSuccess else {
            print("Failed to encode/decode with error code: \(cryptStatus)")
            return nil
        }
        
        return Data(bytes: outputBuffer, count: outputMovedSize)
    }
    
    // Get cached key
    func getCachedKey(from value: String) -> String? {
        guard let index = value.firstIndex(of: "_") else { return nil }
        return String(value[..<index])
    }
    
    // Get cached identifier
    func getCachedIdentifier(from value: String) -> String? {
        guard let index = value.firstIndex(of: "_") else { return nil }
        return String(value[value.index(after: index)...])
    }
    
    // Generate key password (assuming accountID is a property or passed in)
    func generateKeyPassword(accountID: String) -> String {
        return "\(kCRYPT_KEY_PREFIX)\(accountID)\(kCRYPT_KEY_SUFFIX)"
    }
    
    // Encrypt object to base64 string
    func getEncryptedBase64String<T: Codable>(objectToEncrypt: T) -> String? {
        do {
            let encoder = JSONEncoder()
            let dataValue = try encoder.encode(objectToEncrypt)
            
            guard let encryptedData = convertData(dataValue, withOperation: .encrypt) else {
                return nil
            }
            
            return encryptedData.base64EncodedString()
        } catch {
            print("Error encrypting object: \(error)")
            return nil
        }
    }
    
    // Decrypt object from base64 string
    func getDecryptedObject<T: Codable>(encryptedString: String) -> T? {
        guard let dataValue = Data(base64Encoded: encryptedString) else { return nil }
        
        do {
            guard let decryptedData = convertData(dataValue, withOperation: .decrypt) else {
                return nil
            }
            
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: decryptedData)
        } catch {
            print("Error decrypting object: \(error)")
            return nil
        }
    }
    
    // Helper method for encryption/decryption
    private func convertData(_ data: Data, withOperation operation: CryptoOperation) -> Data? {
        // You'll need to implement the key and identifier generation logic here
        // This is a placeholder and should be replaced with your actual key generation method
        let key = "YourDefaultKey"
        let identifier = "YourDefaultIdentifier"
        
        return aes128(
            operation: operation == .encrypt ? CCOperation(kCCEncrypt) : CCOperation(kCCDecrypt),
            key: key,
            identifier: identifier,
            data: data
        )
    }
    
    // Enum to represent crypto operation
    enum CryptoOperation {
        case encrypt
        case decrypt
    }
}

// Constants (you should replace these with your actual values)
private let kCRYPT_KEY_PREFIX = "PREFIX_"
private let kCRYPT_KEY_SUFFIX = "_SUFFIX"
